keps/sig-release/20190316-rebase-images-to-distroless.md

```diff
@@ -0,0 +1,149 @@
+There are 4 main efforts on this KEP.
+
+1. The images are built via bazel. In such case, we will update the bazel BUILD rule to switch to base from distroless image. This method applies for the core containers like kube-apiserver, kube-controller-manager, cloud-controller-manager, kube-scheduler. As for kube-proxy, it is based on debian-iptable which distroless doesn't support iptables yet. Thus, kube-proxy won't be changed.
+2. The images have dependencies that are not supported by distroless. One typical example is the usage of shell. Previously, shell is widely used for redirecting glog output to a certain directory. This use case is no longer needed since we've switched from glog to klog which can accept a flag to specify the log output path. A generic approach is: Remove the dependencies that distroless doesn't support and then rebase the images to distroless (e.g. [issue/1787](https://github.com/kubernetes/autoscaler/issues/1787). Meanwhile, we limit the introduction of new dependencies. (e.g. [pr/74690](https://github.com/kubernetes/kubernetes/pull/74690#discussion_r266037189))
+3. Images based "FROM scratch" is safe to switch to distroless directly.
+4. Images from kubernetes incubator won't be changed directly by this KEP and the release plan is not estimated here. We notify the project OWNERs and we defer to the OWNERs on whether or not those images should be updated.
+
+## Graduation Criteria
+
+This KEP is targeted at v1.15 release. The full list of images switched to distroless will be updated later on.
+
+### Testing
+
+#### Test the Core Images.
+
+Step 1: Use bazel to build the component .tar files.
+```bash
+>> make bazel-release
+```
+output .tar files
+- bazel-bin/build/release-tars/kubernetes-client-linux-amd64.tar.gz
+- bazel-bin/build/release-tars/kubernetes-node-linux-amd64.tar.gz
+- bazel-bin/build/release-tars/kubernetes-server-linux-amd64.tar.gz
+- bazel-bin/build/release-tars/kubernetes-manifests.tar.gz
+- bazel-bin/build/release-tars/kubernetes-test-portable.tar.gz
+- bazel-bin/build/release-tars/kubernetes-test-linux-amd64.tar.gz
+- bazel-bin/build/release-tars/kubernetes-test.tar.gz
+- bazel-bin/build/release-tars/kubernetes.tar.gz
+- bazel-bin/build/release-tars/kubernetes-src.tar.gz
+Step 2: Upload .tar to the bigstore (Google GKE internal test env)
+Step 3: Create a cluster using the uploaded .tar (Google GKE internal test env)
+```bash
+>> gcloud container clusters create --cluster-version 1.14.0-yuwen-distroless-0308115 distroless-test-cluster
+```
+Step 4: Find out the host master VM
+Step 5: Verify the containers running in the master node is created with the distroless image.
+```bash
+>> gcloud compute ssh --project=... --zone=xx
+```
+After ssh into the master VM
+```
+sudo crictl ps
+```
+And see if the target container has shell (should not have) and has ca-certificate (should have).
```

Comment:
you could also just load and examine each of the containers locally?

though as I think you're alluding here, we should also make sure all of the e2e tests are still passing with these changes.

Reply:
Yes, I tried that and found some problems.

The tricky part (which I stuck for a really long time) is that rebasing images from debian to distroless requires removing the shell, which means not only the /bin/sh in the manifest command need to be removed, but the parameters passed into the command need to be reformed as well (as running a executable binary). Local test doesn't give a good coverage since some of the parameters (e.g. kube-apiserver --service-cluster-ip-range) are provided by the runtime env. (I was stuck on the different param value formats when converting bash command (/bin/sh -c) to binary parameters and I only noticed this problem when deploying a real GKE cluster) 

Thanks for your suggestions. Lots of people helped me and gave suggestions until I can depict the workable/non-workable testing path :)